File: ios/Runner/Callkit/Channels/NativeCallingEventChannel.swift
//
//  NativeCallingEventChannel.swift
//  Runner
//
//  Created by Hashim Khan on 24/01/2024.
//

import Flutter
import UIKit

public class NativeCallingEventChannel: NSObject, FlutterStreamHandler {
    
    static var shared: NativeCallingEventChannel  = NativeCallingEventChannel()
    
    var eventSink: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterEventChannel(name: "native_calling_event_channel", binaryMessenger: registrar.messenger())
        channel.setStreamHandler(NativeCallingEventChannel.shared)
    }
    
    public func onListen(withArguments arguments: Any?, eventSink sink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = sink
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    public func sendEvent(event : NativeCallingEvents,data : Any) {
        guard let sink = eventSink else { return }
        sink([
            "event" : event.getName(),
            "data" : data
        ])
    }
}

File: android/app/src/main/kotlin/com/transport_system/ts_driver/native_calling_plugin/NativeCallingEventChannel.kt
package com.transport_system.ts_driver.native_calling_plugin


import io.flutter.plugin.common.EventChannel
import io.flutter.embedding.engine.plugins.FlutterPlugin

class NativeCallingEventChannel : EventChannel.StreamHandler, FlutterPlugin {

    companion object {
        val instance: NativeCallingEventChannel by lazy { NativeCallingEventChannel() }
    }

    private var eventSink: EventChannel.EventSink? = null
    private var eventChannel: EventChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        eventChannel = EventChannel(binding.binaryMessenger, "native_calling_event_channel")
        eventChannel?.setStreamHandler(instance)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        eventChannel?.setStreamHandler(null)
        eventChannel = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }

    fun sendEvent(event: NativeCallingEvents, data: Any?) {
        eventSink?.success(
            mapOf(
                "event" to event.getName(),
                "data" to data
            )
        )
    }
}
